File: leetcode/Python/0082-m-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
Linked List
"""
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head

        prev = dummy
        curt = head

        while curt and curt.next:
            if curt.val == curt.next.val:
                dupVal = curt.val
                while curt and curt.val == dupVal:
                    prev.next = curt.next
                    curt = curt.next
            else:
                prev = prev.next
                curt = curt.next
        return dummy.next

File: leetcode/C++/0082-m-remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* dummy = new ListNode();
        dummy->next = head;
        ListNode* prev = dummy;
        ListNode* cur = head;
        while (cur && cur->next) {
            if (cur->val == cur->next->val) {
                int dupVal = cur->val;
                while (cur && cur->val == dupVal)
                    cur = cur->next;
                prev->next = cur;
            } else {
                prev = prev->next;
                cur = cur->next;
            }
        }
        return dummy->next;
    }
};